text/0000-dyn-sized.md

```diff
@@ -0,0 +1,873 @@
+    pub auto trait DynSized {}
+    ```
+
+    Since `extern type` will not implement any auto traits, this line is enough to distinguish
+    foreign types.
+
+    Whether `DynSized` is an auto trait is an implementation detail, and this RFC does not dictate
+    the choice. However, there may be compilation [performance problems] with auto traits.
+
+## `DynSized` as implied bound
+
+This RFC does not make `DynSized` an implied bound, in order to workaround [RFC issue #2255]. This
+means existing `?Sized` bounds will continue to accept foreign types as input, and we cannot fix the
+bounds of `size_of_val` without breaking backward compatibility. Instead we introduce a lint, and
+make use of epochs to eventually turn the lint into an error. Still, the bounds of `size_of_val`
+will forever be incorrect to support Epoch 2015.
+
+An alternative is just accept that we want `?DynSized`. As stated in [RFC #1993], making `DynSized`
+an implied bound has the following benefits:
+
+* Libraries, including the standard library, can relax `?Sized` to `?DynSized` at their own pace
+    without losing backward compatibility.
+* The standard library can provide the correct bound for all types without the `#[assume_dyn_sized]`
+    hack.
+* `?Sized`/`?DynSized` looks better than `(DynSized + ?Sized)`/`?Sized`.
+
+On the other hand, adding new implied bounds is currently blocked by the compiler [issue #21974].
+Furthermore, concerns are raised about new implied bounds like `?DynSized` and `?Move`:
+
+* Implied bound is a "negative feature", and it is confusing to reason about.
+* Unlike the lints, the user needs to manually evaluate every use of `?Sized` and determine whether
+    moving to `?DynSized` is acceptable.
+
+## Replace `#[assume_dyn_sized]` by `DynSized` itself
+
+Instead of introducing a new attribute for linting, we may make `DynSized` itself serve the purpose
+for linting. We still restrict the bounds for `size_of_val` and friends:
+
+```rust ,ignore
+fn size_of_val<T: DynSized + ?Sized>(val: &T) -> usize;
+//                ^~~~~~~~
+```
+
+However, when we instantiate `size_of_val::<Foreign>`, instead of a type-check error, we emit a lint
+instead. This allows us to describe the parameter using the correct bound without breaking existing
+code.
+
+The drawback is that it makes the trait bound concept very unnatural — the `DynSized` bound is
+written there, but it cannot be used for inference, and violating the bound just causes a warning
+🤔.
+
+## Post-monomorphization error
+
+Instead of introducing `DynSized` or lints, we may simply refuse to compile during monomorphization
+of `size_of_val`, `align_of_val` or struct field using a foreign type. This gives an accurate
+compile-time error and only affects the rare cases where foreign type is actually misused.
+
+Rust currently has no post-monomorphization lints or type errors (except [E0511] or recursion
+overflow). Introducing such errors marks a serious departure of this policy.
```

Comment:
Please don't refer to errors by error codes, or the error message, but rather for the semantic reason for the error.

Also note that other than resource limits, all the post-monomorphization errors come from unstable features (mostly inline assembly and platform intrinsics, both of which shouldn't have stabilized forms that don't move the errors to earlier checks).